frontend/src/components/common/ChangePassword/ChangePasswordModal.tsx

```diff
@@ -0,0 +1,264 @@
+  const [
+    confirmPasswordFeedback,
+    setConfirmPasswordFeedback,
+  ] = React.useState<PasswordFeedback>({ isInvalid: false, errorMessage: "" });
+
+  const [feedback, setFeedback] = React.useState<string>("");
+  const [submitted, setSubmitted] = React.useState<boolean>(false);
+  const [error, setError] = React.useState<boolean>(false);
+
+  const resetFeedback = () => {
+    setFeedback("");
+    setOldPasswordFeedback({ isInvalid: false, errorMessage: "" });
+    setNewPasswordFeedback({ isInvalid: false, errorMessage: "" });
+    setConfirmPasswordFeedback({ isInvalid: false, errorMessage: "" });
+  };
+
+  const reset = () => {
+    setOldPassword("");
+    setNewPassword("");
+    setConfirmPassword("");
+    setSubmitted(false);
+    setError(false);
+    resetFeedback();
+  };
+
+  const submit = () => {
+    resetFeedback();
+
+    if (oldPassword === "" || newPassword === "" || confirmPassword === "") {
+      setOldPasswordFeedback({ isInvalid: true, errorMessage: "Required" });
+      setNewPasswordFeedback({ isInvalid: true, errorMessage: "Required" });
+      setConfirmPasswordFeedback({ isInvalid: true, errorMessage: "Required" });
+      setFeedback("Please enter all fields.");
+    } else if (newPassword !== confirmPassword) {
+      setNewPasswordFeedback({ isInvalid: true, errorMessage: "" });
+      setConfirmPasswordFeedback({
+        isInvalid: true,
+        errorMessage: "Passwords do not match.",
+      });
+    } else if (oldPassword === newPassword) {
+      setOldPasswordFeedback({ isInvalid: true, errorMessage: "" });
+      setNewPasswordFeedback({
+        isInvalid: true,
+        errorMessage: "New password and old password must be different.",
+      });
+    } else {
+      const auth = getAuth(firebaseApp);
+
+      if (authenticatedUser?.email) {
+        // Sign in on the frontend side to validate password
+        signInWithEmailAndPassword(auth, authenticatedUser.email, oldPassword)
+          .then((userCredential) => {
+            // Signed in
+            const { user } = userCredential;
+            updatePassword(user, newPassword)
+              .then(() => {
+                // Update successful
+                setSubmitted(true);
+              })
+              .catch((e) => {
+                // Error occurred
+                if ((e as FirebaseError).code === "auth/weak-password") {
+                  setNewPasswordFeedback({
+                    isInvalid: true,
+                    errorMessage: "Password is too weak.",
+                  });
+                } else {
+                  setSubmitted(true);
+                  setError(true);
+                }
+              });
+          })
+          .catch((e) => {
+            if ((e as FirebaseError).code === "auth/wrong-password") {
+              setOldPasswordFeedback({
+                isInvalid: true,
+                errorMessage: "The password you entered is incorrect.",
+              });
```

Comment:
also a nit, personally i don't even know super well how to do this, but I think syntactically, using Async, Await, is a lot cleaner then using Promises and .then()